compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -691,6 +691,25 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         {
                             return;
                         }
+                        if let ObligationCauseCode::FunctionArgumentObligation {
+                            arg_hir_id,
+                            ..
+                        } = obligation.cause.code()
+                            && let Some(Node::Expr(arg)) = self.tcx.hir().find(*arg_hir_id)
+                            && let arg = arg.peel_borrows()
+                            && let hir::ExprKind::Path(hir::QPath::Resolved(
+                                None,
+                                hir::Path { res: hir::def::Res::Local(hir_id), .. },
+                            )) = arg.kind
+                            && let Some(Node::Pat(pat)) = self.tcx.hir().find(*hir_id)
+                            && let Some(preds) = self.reported_trait_errors.borrow().get(&pat.span)
+                            && preds.contains(&obligation.predicate)
+                            && self.tcx.sess.has_errors().is_some()
```

Comment:
I think we want to move forward with depending on less global state, rather than more.

Reply:
Note that a few lines above we already have a `has_errors` check. When we come up with an alternative for that one, this one can also go away. In this case this is to ensure that we don't unnecessarily silence obligations that happen to be in `reported_trait_errors` without actually having been emitted (I found a couple of examples in the test suite).

Reply:
@compiler-errors let me know if my rationale above is sufficient to land the PR as is, otherwise we can close this PR.

Reply:
Is `tcx.sess.has_errors().is_some()` required when we also check `self.reported_trait_errors.borrow().get(&pat.span)`?

Reply:
The following tests ICE without that check

    [ui] tests/ui/derives/derives-span-Clone-enum-struct-variant.rs
    [ui] tests/ui/derives/derives-span-Clone-enum.rs
    [ui] tests/ui/derives/derives-span-Hash-enum-struct-variant.rs
    [ui] tests/ui/derives/derives-span-PartialOrd-enum-struct-variant.rs
    [ui] tests/ui/derives/derives-span-PartialOrd-enum.rs
    [ui] tests/ui/derives/derives-span-Hash-enum.rs

I think because `report_fulfillment_errors` inserts all the error predicates found in the global `reported_trait_errors`, but it *later* has suppression logic that avoids emitting some diagnostics. I'm sure we could change that logic instead, but will make the PR larger.